### PR TITLE
feat: Implement Decision Tree Classifier from scratch using NumPy

### LINKED ISSUE #7021: Fork, Commit, Merge - Hard Issue (Python)
# Fork, Commit, Merge - Hard Issue (Python)

## Implementing a Decision Tree Classifier from Scratch
*Note: You don't have ask permission to start solving the issue or get assigned, since these issues are supposed to be always open for new contributors. The actions-user bot will reset the file back to previous state for the next contributor after your commit is merged. So you can just simply start working with the issue right away!*

## How to get started

For this task you need to have Python and NumPy installed. Check out [Installing Python](https://github.com/nikohoffren/fork-commit-merge#installing-python) section in README if you need to install Python. If you already have Python installed, you can install NumPy using pip (Python's package installer) with the following terminal command:

```bash
pip install numpy
```

Or, if you're using Python 3 specifically and have both Python 2 and Python 3 installed, you may need to use:

```bash
pip3 install numpy
```

After that you can open the `tasks/python/hard` directory from the root of the project. 
Then open `decision_tree.py` file and start working on your solution!

python-hard

## Description

Implement a Decision Tree Classifier from scratch using Python. Do not use libraries like scikit-learn that have pre-implemented classifiers; instead, use basic libraries like NumPy for numerical calculations.

Objectives:

-   Understand the theory behind Decision Trees.
-   Implement a simple yet functional Decision Tree Classifier.
-   Test the classifier on a dataset.

Use any freely available classification dataset. Ensure the dataset has at least 3 features and at least 2 classes.

Implementation Steps:
-   Write functions to calculate Gini impurity or Information Gain.
-   Implement a class DecisionTree with methods for fitting and predicting.
-   Implement a function to print/visualize the tree.

Testing:
-   Split the dataset into training and test sets and evaluate your classifier.

Acceptance Criteria:

-   Successfully implement a Decision Tree Classifier from scratch.
-   Evaluate the classifier on a dataset, showing reasonable performance metrics (e.g., accuracy, precision, recall).

Resources:

-   [Decision Trees - Wikipedia](https://en.wikipedia.org/wiki/Decision_tree_learning)

## How to run

Make sure you are in the right directory:

```bash
cd tasks/python/hard
```

Execute the following command to run your Python script:

```bash
python decision_tree.py
```

## Expected output

Output should look similar to this:

```bash
The predicted class for the sample [0.4, 0.6] is 0.
```

If the output looks correct, you are ready to make a pull request!

---

To work with this issue, you need to have Python installed to your local machine.
Check out [README.md](https://github.com/nikohoffren/fork-commit-merge/blob/main/README.md) for more instructions of installing Python and how to make a pull request.

Feel free to ask any questions here if you have some problems!

#### Also, kindly give this project a star to enhance its visibility for new developers!

### PR DESCRIPTION
**Pull request by gavin913-lss**
Closes #7021

## Changes
- Implemented DecisionTreeClassifier from scratch using NumPy
- Gini impurity-based splitting with information gain calculation
- Supports configurable max_depth and min_samples_split
- Includes tree visualization via print_tree() method
- Generated synthetic dataset with 3 features and 2 classes
- Full evaluation with accuracy, precision, recall, and F1 score

## Implementation
- DecisionTreeNode: Stores split info (feature, threshold) or leaf value
- DecisionTreeClassifier: Recursive tree building with _build_tree()
- train_test_split: 80/20 split for evaluation
- precision_recall_f1: Binary classification metrics

## Testing
```
python3 tasks/python/hard/decision_tree.py
```

Output achieves ~95% accuracy with proper precision/recall metrics.

The predicted class for the sample [0.4, 0.6, 0.5] is 1.

### CODE REVIEW

**Review by nikohoffren — APPROVED** [review]

**Comment by nikohoffren** [comment]
## Merged

> This is an automated message from Fork, Commit, Merge [BOT].

Thank you for your contribution! Your Pull Request has been merged, and the files have been reset for the next contributor.

### What's next?

If you're looking for more ways to contribute, we invite you to check out our [forkcommitmerge.dev](https://github.com/fork-commit-merge/fork-commit-merge-web) website repository. The repository contains real issues you can help resolve, or you can create some of your own if you notice any bugs or want to suggest new features. You can also check out the [Influences](https://github.com/fork-commit-merge/fork-commit-merge#influences) section in the README to find more projects similar to this one.

Also please leave a star to this project if you found it helpful.

We look forward to seeing your contributions!
